### PR TITLE
bazel: switch to native proto_library/cc_proto_library.

### LINKED ISSUE #1873: Envoy API proto build rules via Bazel native proto support
Bazel has _some_ support for generating protos via its inbuilt `proto_library` and `cc_proto_library` rules today (see https://docs.bazel.build/versions/master/be/protocol-buffer.html). We've so far been using the imported rules for `cc_proto_library` etc from https://github.com/google/protobuf/blob/master/protobuf.bzl, because we wanted a high degree of control over the protobuf compiler and runtime (associated largely with our former approach to external dep build for protobuf prior to #1682).

We may now be able to switch to native support, which has the advantage that the `proto_library` rule provides language neutral support for management of transitive proto dependencies. This is useful when doing things like adding additional proto rules to generate constraint validation code (https://github.com/lyft/protoc-gen-validate) and documentation (#1866), since you can avoid horrible hacks like https://github.com/lyft/protoc-gen-validate/blob/master/bazel/go_proto_library.bzl.diff. This might also make it simpler to add support for additional languages (e.g. Go), since the transitivity management is factored out.

It's a bit unclear to me how mature the support is in native Bazel for the latest protobuf toolchain, so creating this tracking issue to investigate. https://blog.bazel.build/2017/02/27/protocol-buffers.html describes how to get started. One thing we'll need to do is rename our protobuf repo to `@com_google_protobuf`.

@jmillikin-stripe @lizan for comment. @rodaine for PGV applications.

### PR DESCRIPTION
**Pull request by htuch**
Envoy repo side change following https://github.com/envoyproxy/data-plane-api/pull/201.

This allows us to drop the custom patch for protobuf. One not so nice thing is that the native
cc_proto_library doesn't play nice with the hack we had in place to get at generated .pb.h files via
the "Envoy canonical include path", i.e. common/ instead of source/common/. There aren't that many
places this is an issue, but it's a bit uglier.

Also switched back to `-fPIC`; recent versions of Bazel seem to avoid doing multiple redundant builds that used to plague us when PIC was enabled.

Fixes #1873.

Signed-off-by: Harvey Tuch <htuch@google.com>

### CODE REVIEW

**Comment by htuch** [comment]
@lizan @jmillikin-stripe @mattklein123 @kyessenov

File: bazel/unix_cc_configure.bzl [hunk 1]
```diff
@@ -168,7 +168,7 @@ def _crosstool_content(repository_ctx, cc, cpu_value, darwin):
       "builtin_sysroot": "",
       "compiler": escape_string(get_env_var(repository_ctx, "BAZEL_COMPILER", "compiler", False)),
       "host_system_name": escape_string(get_env_var(repository_ctx, "BAZEL_HOST_SYSTEM", "local", False)),
-      "needsPic": False,
+      "needsPic": True,
```
Comment (jmillikin-stripe) [inline bazel/unix_cc_configure.bzl]: Can you recalculate unix_cc_configure.bzl.diff with this change?

File: bazel/repositories.bzl [hunk 2]
```diff
@@ -259,23 +259,30 @@ def com_github_lightstep_lightstep_tracer_cpp(repository = ""):
...
+      name = "com_google_protobuf_cc",
+      strip_prefix = "protobuf-c4f59dcc5c13debc572154c8f636b8a9361aacde",
+      sha256 = "5d4551193416861cb81c3bc0a428f22a6878148c57c31fb6f8f2aa4cf27ff635",
+      url = "https://github.com/google/protobuf/archive/c4f59dcc5c13debc572154c8f636b8a9361aacde.tar.gz",
   )
   native.bind(
-      name="protobuf",
-      actual="@protobuf_bzl//:protobuf",
+      name = "protobuf",
+      actual = "@com_google_protobuf//:protobuf",
```
Comment (lizan) [inline bazel/repositories.bzl]: nit: C++ runtime library should be loaded from `com_google_protobuf_cc`

File: bazel/repositories.bzl [hunk 3]
```diff
@@ -139,19 +139,19 @@ def envoy_api_deps(skip_targets):
...
+        actual = "@googleapis//:http_api_protos_lib",
     )
 
-def envoy_dependencies(path = "@envoy_deps//", skip_protobuf_bzl = False, skip_targets = [],
+def envoy_dependencies(path = "@envoy_deps//", skip_com_google_protobuf = False, skip_targets = [],
                        repository = ""):
     native.bind(
         name = "cc_wkt_protos",
-        actual = "@protobuf_bzl//:cc_wkt_protos",
+        actual = "@com_google_protobuf//:cc_wkt_protos",
```
Comment (lizan) [inline bazel/repositories.bzl]: `com_google_protobuf_cc`

File: bazel/repositories.bzl [hunk 4]
```diff
@@ -139,19 +139,19 @@ def envoy_api_deps(skip_targets):
...
     )
     native.bind(
         name = "cc_wkt_protos_genproto",
-        actual = "@protobuf_bzl//:cc_wkt_protos_genproto",
+        actual = "@com_google_protobuf//:cc_wkt_protos_genproto",
```
Comment (lizan) [inline bazel/repositories.bzl]: `com_google_protobuf_cc`

File: bazel/repositories.bzl [hunk 5]
```diff
@@ -259,23 +259,30 @@ def com_github_lightstep_lightstep_tracer_cpp(repository = ""):
...
-      name = "protobuf_bzl",
-      urls = [
-          "https://github.com/google/protobuf/releases/download/v3.4.0/protobuf-cpp-3.4.0.tar.gz",
-      ],
-      sha256 = "71434f6f836a1e479c44008bb033b2a8b2560ff539374dcdefb126be739e1635",
-      strip_prefix = "protobuf-3.4.0",
-      patches = [
-          repository + "//bazel/external:protobuf-memory-errors.patch",
-      ],
+def com_google_protobuf(repository = ""):
```
Comment (lizan) [inline bazel/repositories.bzl]: unused parameter `repository`?

File: bazel/external/lightstep.BUILD [hunk 6]
```diff
@@ -17,7 +17,7 @@ cc_library(
     hdrs = [":" + hdr for hdr in _PREFIX_HDRS],
     includes = ["_prefix/include"],
     visibility = ["//visibility:public"],
-    deps = ["@protobuf_bzl//:protobuf"],
+    deps = ["@com_google_protobuf//:protobuf"],
```
Comment (lizan) [inline bazel/external/lightstep.BUILD]: Can those be `//external` ?

Comment (htuch) [reply bazel/external/lightstep.BUILD]: We're moving away from bind now that we're switching to the Bazel dictated `@com_google_protobuf`.

Comment (lizan) [reply bazel/external/lightstep.BUILD]: OK then load those from `@com_google_protobuf_cc` if it is C++ specific.

**Comment by jmillikin-stripe** [comment]
LGTM

**Pull request merged**